src/core/qgsdiagramrenderer.cpp

```diff
@@ -804,6 +837,255 @@ void QgsLinearlyInterpolatedDiagramRenderer::writeXml( QDomElement &layerElem, Q
+{
+  if ( !mDiagram )
+  {
+    return;
+  }
+
+  QPointF newPos = pos; // Each subdiagram will have its own newPos
+  QList< QgsDiagramRenderer * > stackedRenderers = renderers();
+  for ( const auto &stackedRenderer : std::as_const( stackedRenderers ) )
+  {
+    if ( stackedRenderer->rendererName() == QStringLiteral( "Stacked" ) )
+    {
+      stackedRenderer->renderDiagram( feature, c, newPos, properties );
+      continue;
+    }
+
+    QgsDiagramSettings s;
+    if ( !stackedRenderer->diagramSettings( feature, c, s ) )
+    {
+      return;
+    }
+
+    if ( properties.hasActiveProperties() )
+    {
+      c.expressionContext().setOriginalValueVariable( QgsColorUtils::colorToString( s.backgroundColor ) );
+      s.backgroundColor = properties.valueAsColor( QgsDiagramLayerSettings::Property::BackgroundColor, c.expressionContext(), s.backgroundColor );
+      c.expressionContext().setOriginalValueVariable( QgsColorUtils::colorToString( s.penColor ) );
+      s.penColor = properties.valueAsColor( QgsDiagramLayerSettings::Property::StrokeColor, c.expressionContext(), s.penColor );
+      c.expressionContext().setOriginalValueVariable( s.penWidth );
+      s.penWidth = properties.valueAsDouble( QgsDiagramLayerSettings::Property::StrokeWidth, c.expressionContext(), s.penWidth );
+      c.expressionContext().setOriginalValueVariable( s.rotationOffset );
+      s.rotationOffset = properties.valueAsDouble( QgsDiagramLayerSettings::Property::StartAngle, c.expressionContext(), s.rotationOffset );
+    }
+
+    QgsPaintEffect *effect = s.paintEffect();
+    std::unique_ptr< QgsEffectPainter > effectPainter;
+    if ( effect && effect->enabled() )
+    {
+      effectPainter = std::make_unique< QgsEffectPainter >( c, effect );
+    }
+
+    stackedRenderer->diagram()->renderDiagram( feature, c, s, newPos );
+    QgsStackedDiagram *stackedDiagram = dynamic_cast< QgsStackedDiagram *>( mDiagram.get() );
+    stackedDiagram->subDiagramPosition( newPos, c, mSettings, s );
+  }
+}
+
+bool QgsStackedDiagramRenderer::diagramSettings( const QgsFeature &feature, const QgsRenderContext &c, QgsDiagramSettings &s ) const
+{
+  Q_UNUSED( feature )
+  Q_UNUSED( c )
+  Q_UNUSED( s )
+  return false;
+}
+
+QSizeF QgsStackedDiagramRenderer::diagramSize( const QgsFeature &feature, const QgsRenderContext &c ) const
+{
+  Q_UNUSED( feature )
+  Q_UNUSED( c )
+  return QSizeF( 0, 0 );
+}
+
+QList<QgsDiagramSettings> QgsStackedDiagramRenderer::diagramSettings() const
+{
+  QList<QgsDiagramSettings> settingsList;
+  settingsList.push_back( mSettings );
+  return settingsList;
+}
+
+QList<QString> QgsStackedDiagramRenderer::diagramAttributes() const
+{
+  return mSettings.categoryAttributes;
+}
+
+QList< QgsLayerTreeModelLegendNode * > QgsStackedDiagramRenderer::legendItems( QgsLayerTreeLayer *nodeLayer ) const
+{
+  QList< QgsLayerTreeModelLegendNode * > nodes;
+  for ( int i = 0; i < rendererCount(); i++ )
```

Comment:
Any reason not to use a range based for loop here?

`for( QgsDiagramRenderer *renderer : mDiagramRenderers )`?

Reply:
Not really @m-kuhn, I could adjust it.

Reply:
They are normally preferable

Reply:
Addressed in https://github.com/qgis/QGIS/commit/f81df2876c9874f251eb01d7551f443fa37e2df3#diff-c2267f7efae558b8c9bc33e2e751364580eb0b84a7b9b709d3ac815b601bf8eaR977

Is this form also welcome?

`for ( const auto &renderer : std::as_const( mDiagramRenderers ) )`